### PR TITLE
Port VSCode's glob-to-regexp conversion

### LINKED ISSUE #1979: FileSystemWatcher.globPattern support incomplete
**Describe the bug**

In https://github.com/golang/tools/commit/fd294ab11aed8d198f7a12a86dfe0a970fdab143 and after, gopls changed to using a [DidChangeWatchedFiles globPattern](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_didChangeWatchedFiles) which uses groups:

```
/some-path/**/*.{go,mod,sum}
```

(the pattern was changed further in https://github.com/golang/tools/commit/102e7d35703151d99b09a7158e6fc667492075b5)

As part of [debugging a fix for a gopls issue](https://github.com/golang/go/issues/40456#issuecomment-666455170), I think I discovered lsp-mode does not support groups in globPatterns.

I added some quick debugging to lsp-mode.el around [here](https://github.com/emacs-lsp/lsp-mode/blob/0131f1f4c9e7823955c3b1f80881fcd7c2eaa551/lsp-mode.el#L3163) and saw that `glob-pattern` doesn't seem to do proper expansion for groups:

```
lsp file check changed=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456/go.mod root-folder=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456 glob-pattern=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456/**/*.{go,mod,sum} glob-regex=\`/Users/dan\.peterson/Projects/go/src/github\.com/danp/go-40456/.*.*/.*\.{go,mod,sum}\' relative=go.mod changed-match=nil relative-matched=nil
```

The debug logging line looks like:

```elisp
(lsp-log "lsp file check changed=%s root-folder=%s glob-pattern=%s glob-regex=%s relative=%s changed-match=%s relative-matched=%s" changed-file root-folder glob-pattern glob-regex (f-relative changed-file root-folder) (string-match glob-regex changed-file) (string-match glob-regex (f-relative changed-file root-folder)))
```

In addition to groups, it's possible eshell-glob-regexp doesn't fully support the complete syntax that FileSystemWatcher.globPattern allows.

**To Reproduce**
Open https://github.com/danp/go-40456 with lsp-mode and gopls at or after https://github.com/golang/tools/commit/102e7d35703151d99b09a7158e6fc667492075b5.

**Expected behavior**
Especially when trying the fix for https://github.com/golang/go/issues/40456 I expected to see results of gopls being notified of watched files changing.

**Which Language Server did you use**
gopls

**OS**
macOS

### PR DESCRIPTION
**Pull request by leungbk**
Closes #1979.

### CODE REVIEW

File: lsp-clients.el [hunk 1]
```diff
@@ -979,7 +979,7 @@ responsiveness at the cost of possible stability issues."
 
 (defun parse-rf-language-server-globs-to-regex (vector)
   "Converts vector with globs to regex"
```
Comment (yyoncho) [inline lsp-clients.el:981]: please, remove the require about eshell.

**Comment by leungbk** [comment]
@danp does this resolve your issue? I'm not knowledgeable with go, but I still seem to be getting `didChangeWatchedFiles` only for `.go` files, even when I turn on `lsp-mode` inside `go.mod`.

**Comment by danp** [comment]
@leungbk Thanks! I tried patching my local lsp-mode with these changes and added my debugging from #1979. Here's what I see when changing between the main and change branches:

```
lsp file check changed=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456/go.mod root-folder=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456 glob-pattern=file:///Users/dan.peterson/Projects/go/src/github.com/danp/go-40456/**/*.{go,mod,sum} glob-regex=\`file:[/\\][/\\][/\\]Users[/\\]dan\.peterson[/\\]Projects[/\\]go[/\\]src[/\\]github\.com[/\\]danp[/\\]go-40456[/\\]\(?:[/\\]\|[^/\\]+[/\\]\|[/\\][^/\\]+\)*?[^/\\]*?\.\(?:go\|mod\|sum\)\' relative=go.mod changed-match=nil relative-matched=nil
lsp file check changed=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456/go.sum root-folder=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456 glob-pattern=file:///Users/dan.peterson/Projects/go/src/github.com/danp/go-40456/**/*.{go,mod,sum} glob-regex=\`file:[/\\][/\\][/\\]Users[/\\]dan\.peterson[/\\]Projects[/\\]go[/\\]src[/\\]github\.com[/\\]danp[/\\]go-40456[/\\]\(?:[/\\]\|[^/\\]+[/\\]\|[/\\][^/\\]+\)*?[^/\\]*?\.\(?:go\|mod\|sum\)\' relative=go.sum changed-match=nil relative-matched=nil
lsp file check changed=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456/main.go root-folder=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456 glob-pattern=file:///Users/dan.peterson/Projects/go/src/github.com/danp/go-40456/**/*.{go,mod,sum} glob-regex=\`file:[/\\][/\\][/\\]Users[/\\]dan\.peterson[/\\]Projects[/\\]go[/\\]src[/\\]github\.com[/\\]danp[/\\]go-40456[/\\]\(?:[/\\]\|[^/\\]+[/\\]\|[/\\][^/\\]+\)*?[^/\\]*?\.\(?:go\|mod\|sum\)\' relative=main.go changed-match=nil relative-matched=nil
```

Seems to be on the right track but maybe some confusion with the leading `file://` in the glob pattern? It looks like that was changed in https://github.com/golang/tools/commit/383b97c0b5ea2532304cf9403f537b784b9871f2. If I try using a released version of gopls before that change (https://github.com/golang/tools/releases/tag/gopls%2Fv0.4.4), I see matches as expected:

```
lsp file check changed=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456/go.sum root-folder=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456 glob-pattern=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456/**/*.{go,mod,sum} glob-regex=\`[/\\]Users[/\\]dan\.peterson[/\\]Projects[/\\]go[/\\]src[/\\]github\.com[/\\]danp[/\\]go-40456[/\\]\(?:[/\\]\|[^/\\]+[/\\]\|[/\\][^/\\]+\)*?[^/\\]*?\.\(?:go\|mod\|sum\)\' relative=go.sum changed-match=0 relative-matched=nil
lsp file check changed=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456/main.go root-folder=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456 glob-pattern=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456/**/*.{go,mod,sum} glob-regex=\`[/\\]Users[/\\]dan\.peterson[/\\]Projects[/\\]go[/\\]src[/\\]github\.com[/\\]danp[/\\]go-40456[/\\]\(?:[/\\]\|[^/\\]+[/\\]\|[/\\][^/\\]+\)*?[^/\\]*?\.\(?:go\|mod\|sum\)\' relative=main.go changed-match=0 relative-matched=nil
lsp file check changed=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456/main.go root-folder=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456 glob-pattern=/Users/dan.peterson/Projects/go/src/github.com/danp/go-40456/**/*.{go,mod,sum} glob-regex=\`[/\\]Users[/\\]dan\.peterson[/\\]Projects[/\\]go[/\\]src[/\\]github\.com[/\\]danp[/\\]go-40456[/\\]\(?:[/\\]\|[^/\\]+[/\\]\|[/\\][^/\\]+\)*?[^/\\]*?\.\(?:go\|mod\|sum\)\' relative=main.go changed-match=0 relative-matched=nil
```

I'm not sure if glob patterns should be prefixed with `file://` or not when sent by the client.

**Comment by leungbk** [comment]
> I'm not sure if glob patterns should be prefixed with `file://` or not when sent by the client.

OK, I've resubmitted with `file://` trimmed off.

Do you think we should include the corresponding watcher in `lsp-go`, as with https://github.com/emacs-lsp/lsp-mode/blob/efe4847e9c7710b24f8ab9158ecc2c275bede356/lsp-eslint.el#L258-L263, or should we leave it to the user to make their own file-watchers?

**Comment by danp** [comment]
It turns out the `file://` prefix from gopls was not intentional, will be fixed by https://github.com/golang/go/issues/40661. The LSP spec is unclear but it seems the glob pattern is not meant to be URI-based. So should be okay to remove the trimming of `file://` from the start.

Re setting up watchers for Go, I'd say leave it to the language server until a particular need comes up? At least gopls seems fine to set things up itself.

**Comment by leungbk** [comment]
@danp OK, I removed the trimming of `file://`.

@yyoncho The latest Emacs snapshot uses a new version of Eldoc. CI fails because `eldoc-message` is now an obsolete function, and the new Eldoc recommends that we instead use `eldoc-documentation-functions`, a hook variable not available unless we use the latest Eldoc from Melpa. Should we require this newer version as a dependency, or should we simply allow the snapshot to fail without holding up the other CI builds? The new Eldoc requires Emacs >= 26.3.

**Comment by yyoncho** [comment]
@leungbk we should use `ignore-warning` locally to fix the build until we migrate to new eldoc. The migration should happen only if that new eldoc works with emacs 26.x.

**Comment by leungbk** [comment]
> @leungbk we should use `ignore-warning` locally to fix the build until we migrate to new eldoc. The migration should happen only if that new eldoc works with emacs 26.x.

OK, I wrapped `project-roots` and `eldoc-message`. Ready for re-review.

**Pull request merged**